File: deepeval/inspect/loader.py
"""Load saved test runs into nested `Trace` / `BaseSpan` view models.

Sources are either a `test_run_*.json` file or a SQLite store addressed as
`<path>/deepeval.db#<run_id>` (`#<run_id>` optional: latest run). Both
resolve to the same `TestRun` dict, so everything below `_load_payload`
is shared.

The on-disk shape is `TraceApi`: five flat span buckets linked via
`parentUuid`. The loader pops the buckets, validates each span dict
into a single `BaseSpan` class (the `.type` field is the discriminator),
and wires up `children` / `root_spans`.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict, List, Optional

from deepeval.inspect.types import BaseSpan, Trace


class InspectLoadError(Exception):
    """File unreadable or top-level JSON malformed."""


class NoTracesError(InspectLoadError):
    """File parsed but contains zero traces."""


_SPAN_BUCKETS: List[str] = [
    "baseSpans",
    "agentSpans",
    "llmSpans",
    "retrieverSpans",
    "toolSpans",
]


def find_latest_test_run(folder: str | Path) -> Path:
    """Most recently modified `test_run_*.json` under `folder`.

    Sorted by mtime (not filename) so a manually-copied file with a
    stale timestamp in its name still ranks correctly.
    """

    folder_path = Path(folder)
    if not folder_path.is_dir():
        raise FileNotFoundError(
            f"Results folder not found: {folder_path}. "
            "Pass `results_folder=...` to `DisplayConfig(...)` or set "
            "the `DEEPEVAL_RESULTS_FOLDER` env var."
        )

    candidates = sorted(
        folder_path.glob("test_run_*.json"),
        key=lambda p: p.stat().st_mtime,
        reverse=True,
    )
    if not candidates:
        raise FileNotFoundError(
            f"No test_run_*.json files found in {folder_path}."
        )
    return candidates[0]


def _load_payload(source: str | Path) -> Dict[str, Any]:
    """Resolve `source` (JSON file or `<db>#<run_id>`) to a `TestRun` dict."""
    from deepeval import sqlite_store

    parsed = sqlite_store.parse_source(source)
    if parsed is not None:
        db_path, run_id = parsed
        try:
            _, data = sqlite_store.load_test_run_payload(db_path, run_id)
        except (FileNotFoundError, LookupError) as e:
            raise InspectLoadError(str(e)) from e
        except Exception as e:  # sqlite3.Error, json errors, ...
            raise InspectLoadError(
                f"Failed to read test run from {source}: {e}"
            ) from e
    else:
        p = Path(source)
        try:
            with open(p, "r", encoding="utf-8") as f:
                data = json.load(f)
        except (OSError, json.JSONDecodeError) as e:
            raise InspectLoadError(
                f"Failed to read test run from {p}: {e}"
            ) from e

    if not isinstance(data, dict):
        raise InspectLoadError(
            f"Expected the test run in {source} to be an object; "
            f"got {type(data).__name__}."
        )
    return data


def traces_from_test_run_dict(
    data: Dict[str, Any], source: str = "test run"
) -> List[Trace]:
    traces: List[Trace] = []
    for case in data.get("testCases", []):
        trace_dict = case.get("trace") if isinstance(case, dict) else None
        if trace_dict:
            traces.append(_parse_trace(trace_dict))

    if not traces:
        raise NoTracesError(
            f"Unable to inspect {source}: this test run has no traces.\n"
            "See stored runs with `deepeval inspect --list`, or "
            "`deepeval inspect --help` for options."
        )
    return traces


def load_test_run(path: str | Path) -> List[Trace]:
    data = _load_payload(path)
    return traces_from_test_run_dict(data, source=str(path))


def _parse_trace(trace_dict: Dict[str, Any]) -> Trace:
    # Pop the bucket keys before validating so the residual dict is
    # clean trace-level data; spans are validated and linked separately.
    trace_dict = dict(trace_dict)
    span_dicts: List[Dict[str, Any]] = []
    for bucket_name in _SPAN_BUCKETS:
        for span_dict in trace_dict.pop(bucket_name, None) or []:
            if isinstance(span_dict, dict):
                span_dicts.append(span_dict)

    spans = _build_span_tree(span_dicts)
    roots = [s for s in spans.values() if not _has_known_parent(s, spans)]
    roots.sort(key=lambda s: s.start_time or "")

    trace = Trace.model_validate(trace_dict)
    trace.root_spans = roots
    return trace


def _build_span_tree(
    span_dicts: List[Dict[str, Any]],
) -> Dict[str, BaseSpan]:
    by_uuid: Dict[str, BaseSpan] = {}
    for span_dict in span_dicts:
        span = BaseSpan.model_validate(span_dict)
        # On UUID collision keep the first occurrence; silently swapping
        # would be the wrong fallback for malformed input.
        if span.uuid not in by_uuid:
            by_uuid[span.uuid] = span

    for span in by_uuid.values():
        parent = by_uuid.get(span.parent_uuid) if span.parent_uuid else None
        if parent is not None:
            parent.children.append(span)

    for span in by_uuid.values():
        span.children.sort(key=lambda c: c.start_time or "")

    return by_uuid


def _has_known_parent(span: BaseSpan, by_uuid: Dict[str, BaseSpan]) -> bool:
    return bool(span.parent_uuid) and span.parent_uuid in by_uuid


def run_id_from_path(path: str | Path) -> str:
    from deepeval import sqlite_store

    parsed = sqlite_store.parse_source(path)
    if parsed is not None:
        db_path, run_id = parsed
        return (
            f"{db_path.name}#{run_id}" if run_id is not None else db_path.name
        )
    return Path(path).stem


def summarize_test_run(path: str | Path) -> Optional[Dict[str, Any]]:
    """Run-level pass/fail + duration counts for the header bar.

    Returns `None` if the source can't be read — the header then falls
    back to showing just the run id and trace count.
    """

    try:
        data = _load_payload(path)
    except InspectLoadError:
        return None

    return {
        "test_passed": data.get("testPassed"),
        "test_failed": data.get("testFailed"),
        "run_duration": data.get("runDuration"),
        "evaluation_cost": data.get("evaluationCost"),
    }

File: typescript/src/inspect/loader.ts
// Reads persisted test runs into nested trace trees. On disk a trace is a
// `TraceApi`: five flat span buckets whose members point at each other through
// `parentUuid`.
//
// A source is either a `test_run_*.json` file or a SQLite store addressed as
// `<path>/deepeval.db#<run_id>` (`#<run_id>` optional: latest run). Both
// resolve to the same run object, so everything below `readPayload` is shared.

import * as fs from "fs";
import * as path from "path";
import {
  HIDDEN_DIR,
  LATEST_TEST_RUN_FILE,
  DEEPEVAL_LOCAL_STORE,
  DEEPEVAL_RESULTS_FOLDER,
} from "@/constants";
import {
  LOCAL_STORE_SQLITE,
  normalizeLocalStoreMode,
} from "@/sqlite-store/mode";
import {
  DB_FILENAME,
  TestRunNotFoundError,
  formatSource,
  loadTestRunPayload,
  parseSource,
  resolveDbPath,
} from "@/sqlite-store/store";
import type { BaseApiSpan } from "@/tracing/api";
import type { InspectSpan, InspectTrace, RunSummary } from "@/inspect/model";

export class InspectLoadError extends Error {}
export class NoTracesError extends InspectLoadError {}

const SPAN_BUCKETS = [
  "baseSpans",
  "agentSpans",
  "llmSpans",
  "retrieverSpans",
  "toolSpans",
] as const;

type Json = Record<string, unknown>;

function isRecord(value: unknown): value is Json {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/** Ranked by mtime, so a run copied in under a stale name still sorts first. */
export function findLatestTestRun(folder: string): string {
  if (!fs.existsSync(folder) || !fs.statSync(folder).isDirectory()) {
    throw new InspectLoadError(`Results folder not found: ${folder}`);
  }

  const candidates = fs
    .readdirSync(folder)
    .filter((f) => f.startsWith("test_run_") && f.endsWith(".json"))
    .map((f) => path.join(folder, f))
    .map((f) => ({ file: f, mtime: fs.statSync(f).mtimeMs }))
    .sort((a, b) => b.mtime - a.mtime);

  if (candidates.length === 0) {
    throw new InspectLoadError(`No test_run_*.json files found in ${folder}.`);
  }
  return candidates[0]!.file;
}

/** Prefer `deepeval.db` in sqlite mode (read without the Node gate: the DB may
 *  have been written elsewhere) or when a run id was requested. */
function sqliteModeRequested(): boolean {
  return (
    normalizeLocalStoreMode(process.env[DEEPEVAL_LOCAL_STORE]) ===
    LOCAL_STORE_SQLITE
  );
}

function dbSource(dbPath: string, runId?: number | null): string {
  return runId === undefined || runId === null
    ? dbPath
    : formatSource(dbPath, runId);
}

/**
 * Latest run inside `folder`: the newest `test_run_*.json`, or the
 * `deepeval.db` inside it when sqlite mode is on, a run id was asked for, or
 * there are no JSON exports to fall back on.
 */
function findLatestInFolder(folder: string, runId?: number | null): string {
  const dbPath = path.join(folder, DB_FILENAME);
  const hasDb = fs.existsSync(dbPath);
  if (hasDb && (runId != null || sqliteModeRequested())) {
    return dbSource(dbPath, runId);
  }
  try {
    return findLatestTestRun(folder);
  } catch (e) {
    if (hasDb) return dbSource(dbPath, runId);
    throw e;
  }
}

export interface ResolveInspectOptions {
  /** Open this run id from a SQLite store instead of the latest. */
  runId?: number | null;
}

export function resolveInspectTarget(
  target?: string,
  folderOption?: string,
  { runId }: ResolveInspectOptions = {},
): string {
  if (target) {
    const parsed = parseSource(target);
    if (parsed) {
      if (!fs.existsSync(parsed.dbPath)) {
        throw new InspectLoadError(`SQLite store not found: ${parsed.dbPath}`);
      }
      return dbSource(parsed.dbPath, runId ?? parsed.runId);
    }
    if (!fs.existsSync(target)) {
      throw new InspectLoadError(`No such file or directory: ${target}`);
    }
    return fs.statSync(target).isDirectory()
      ? findLatestInFolder(target, runId)
      : target;
  }

  const folder = folderOption || process.env[DEEPEVAL_RESULTS_FOLDER];
  if (folder && folder.trim() !== "") {
    return findLatestInFolder(folder.trim(), runId);
  }

  const defaultDb = resolveDbPath(undefined);
  const hasDefaultDb = fs.existsSync(defaultDb);
  if (hasDefaultDb && (runId != null || sqliteModeRequested())) {
    return dbSource(defaultDb, runId);
  }

  const rolling = path.join(process.cwd(), HIDDEN_DIR, LATEST_TEST_RUN_FILE);
  if (fs.existsSync(rolling)) return rolling;

  // A DB written earlier in sqlite mode is still the best fallback even if
  // the current process is back in json mode.
  if (hasDefaultDb) return dbSource(defaultDb, runId);

  throw new InspectLoadError(
    `No test run found. Expected ${path.join(HIDDEN_DIR, LATEST_TEST_RUN_FILE)} ` +
      "in the current directory. Run an evaluation first, pass a path " +
      "explicitly, or set DEEPEVAL_RESULTS_FOLDER.",
  );
}

function readJson(file: string): Json {
  let raw: string;
  try {
    raw = fs.readFileSync(file, "utf-8");
  } catch (e) {
    throw new InspectLoadError(
      `Failed to read test run from ${file}: ${(e as Error).message}`,
    );
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (e) {
    throw new InspectLoadError(
      `${file} is not valid JSON: ${(e as Error).message}`,
    );
  }

  if (!isRecord(parsed)) {
    throw new InspectLoadError(
      `Expected the top-level JSON in ${file} to be an object.`,
    );
  }
  return parsed;
}

/** Resolve a source (JSON file or `<db>#<run_id>`) to a run object. */
function readPayload(source: string): Json {
  const parsed = parseSource(source);
  if (!parsed) return readJson(source);
  try {
    const { payload } = loadTestRunPayload(parsed.dbPath, parsed.runId);
    if (!isRecord(payload)) {
      throw new InspectLoadError(
        `Expected the test run in ${source} to be an object.`,
      );
    }
    return payload;
  } catch (e) {
    if (e instanceof InspectLoadError) throw e;
    if (e instanceof TestRunNotFoundError) {
      throw new InspectLoadError(e.message);
    }
    throw new InspectLoadError(
      `Failed to read test run from ${source}: ${(e as Error).message}`,
    );
  }
}

/** Header label: `test_run_x` for files, `deepeval.db#3` for stored runs. */
export function runIdFromSource(source: string): string {
  const parsed = parseSource(source);
  if (parsed) {
    const base = path.basename(parsed.dbPath);
    return parsed.runId === null ? base : `${base}#${parsed.runId}`;
  }
  return path.parse(source).name;
}

/** TypeScript nests the trace under `cases[].entry`; Python puts it on `testCases[]`. */
function* iterateCases(
  data: Json,
): Generator<{ trace: Json; name?: string; passed?: boolean }> {
  for (const group of [data.cases, data.testCases]) {
    if (!Array.isArray(group)) continue;
    for (const item of group) {
      if (!isRecord(item)) continue;
      const holder = isRecord(item.entry) ? item.entry : item;
      if (isRecord(holder.trace)) {
        yield {
          trace: holder.trace,
          name: typeof holder.name === "string" ? holder.name : undefined,
          passed:
            typeof holder.success === "boolean" ? holder.success : undefined,
        };
      }
    }
  }
}

/** Also the compile-time check that `InspectSpan` matches what the tracer writes. */
function toInspectSpan(span: BaseApiSpan): InspectSpan {
  return { ...span, children: [] };
}

function buildSpanTree(spans: Json[]): Map<string, InspectSpan> {
  const byUuid = new Map<string, InspectSpan>();

  for (const raw of spans) {
    const uuid = typeof raw.uuid === "string" ? raw.uuid : undefined;
    if (!uuid || byUuid.has(uuid)) continue;
    byUuid.set(uuid, toInspectSpan(raw as unknown as BaseApiSpan));
  }

  for (const span of byUuid.values()) {
    const parent = span.parentUuid ? byUuid.get(span.parentUuid) : undefined;
    if (parent) parent.children.push(span);
  }

  const byStartTime = (a: InspectSpan, b: InspectSpan) =>
    (a.startTime ?? "").localeCompare(b.startTime ?? "");
  for (const span of byUuid.values()) span.children.sort(byStartTime);

  return byUuid;
}

function parseTrace(raw: Json): InspectTrace {
  const rest: Json = { ...raw };
  const spans: Json[] = [];

  for (const bucket of SPAN_BUCKETS) {
    const value = rest[bucket];
    delete rest[bucket];
    if (!Array.isArray(value)) continue;
    for (const span of value) if (isRecord(span)) spans.push(span);
  }

  const byUuid = buildSpanTree(spans);
  const roots = [...byUuid.values()]
    .filter((s) => !s.parentUuid || !byUuid.has(s.parentUuid))
    .sort((a, b) => (a.startTime ?? "").localeCompare(b.startTime ?? ""));

  return { ...(rest as unknown as InspectTrace), rootSpans: roots };
}

export function loadTestRun(file: string): InspectTrace[] {
  const data = readPayload(file);
  const traces: InspectTrace[] = [];

  for (const { trace, name, passed } of iterateCases(data)) {
    traces.push({ ...parseTrace(trace), caseName: name, casePassed: passed });
  }

  if (traces.length === 0) {
    throw new NoTracesError(
      `Unable to inspect ${file}: this test run has no traces.\n` +
        "See stored runs with `npx deepeval inspect --list`, or " +
        "`npx deepeval inspect --help` for options.",
    );
  }
  return traces;
}

export function summarizeTestRun(file: string): RunSummary | null {
  try {
    const data = readPayload(file);
    const num = (key: string): number | undefined =>
      typeof data[key] === "number" ? (data[key] as number) : undefined;
    return {
      testPassed: num("testPassed"),
      testFailed: num("testFailed"),
      runDuration: num("runDuration"),
      evaluationCost: num("evaluationCost"),
    };
  } catch {
    return null;
  }
}
